lib/sparse/src/index/search_context.rs

```diff
@@ -0,0 +1,454 @@
+        // Initialize min record id with max value
+        let mut min_record_id = u32::MAX;
+        // Indicates that posting iterators are not empty
+        let mut found = false;
+
+        // Iterate first time to find min record id at the head of the posting lists
+        for posting_iterator in to_inspect.iter() {
+            if let Some(element) = posting_iterator.posting_list_iterator.peek() {
+                found = true;
+                if element.id < min_record_id {
+                    min_record_id = element.id;
+                }
+            }
+        }
+
+        if !found {
+            return None;
+        }
+        Some(min_record_id)
+    }
+
+    /// Make sure the longest posting list is at the head of the posting list iterators
+    pub fn sort_posting_lists_by_len(&mut self) {
+        // decreasing order
+        self.postings_iterators.sort_by(|a, b| {
+            b.posting_list_iterator
+                .len_to_end()
+                .cmp(&a.posting_list_iterator.len_to_end())
+        });
+    }
+
+    pub fn search(&mut self) -> Vec<ScoredCandidate> {
+        while let Some(candidate) = self.advance() {
+            // push candidate to result queue
+            self.result_queue.push(candidate);
+
+            // we potentially have enough results to prune low performing posting lists
+            if self.result_queue.len() == self.top {
+                // current min score
+                let min_score = self.result_queue.top().unwrap().score;
+
+                // sort posting lists by length to try to prune the longest one
+                self.sort_posting_lists_by_len();
+
+                self.prune_longest_posting_list(min_score);
+            }
+        }
+        // posting iterators exhausted, return result queue
+        let queue = std::mem::take(&mut self.result_queue);
+        queue.into_vec()
+    }
+
+    /// Prune posting lists that cannot possibly contribute to the top results
+    /// Assumes longest posting list is at the head of the posting list iterators
+    /// Returns true if the longest posting list was pruned
+    pub fn prune_longest_posting_list(&mut self, min_score: f32) -> bool {
+        // compute skip target before acquiring mutable reference to posting list iterator
+        let skip_to = if self.postings_iterators.len() == 1 {
+            // if there is only one posting list iterator, we can skip to the end
+            None
+        } else {
+            // otherwise, we skip to the next min elements in the remaining posting list
+            Self::next_min(&self.postings_iterators[1..])
+        };
+
+        let posting_iterator = &mut self.postings_iterators[0];
+        let posting_query_offset = posting_iterator.query_weight_offset;
+        if let Some(element) = posting_iterator.posting_list_iterator.peek() {
+            let max_weight_from_list = element.weight.max(element.max_next_weight);
```

Comment:
I can see the potential memory overhead of repeating values.

We really want to avoid traversing the posting list more than necessary.

One solution is to introduce an additional datastructure per posting list to track the max values in a more compact fashion.

It is a potential improvement, not sure I want to address it now given the extra complexity :thinking: